frida_tools/tracer.py

```diff
@@ -678,7 +706,11 @@ def _create_trace_script(self):
         return {};
     }
 }
-"""
+
+function initializeSession() {
+    eval("%(init_code)s");
```

Comment:
This breaks if a session script contains double quotes. It would be good to pass it in through RPC instead. (Like we discussed on the other PR.)

Reply:
I need your help here.

I modified the code to execute all --init-session JS files in the `rpc.exports: init` function.  Unfortunately, when running the code, the `init` function is not being invoked.  I will commit the non-working code for your examination.

Here is the relevant `tracer.py` code:

```
    def _create_trace_script(self):
        init_session_code = ""
        for path in self._init_session_paths:
            with open(path, 'rt') as fo:
                lines = fo.readlines()

            init_session_code += "\n".join(lines)
            init_session_code += "\n"

        return """'use strict';

var started = Date.now();
var handlers = {};
var state = {};
var pending = [];
var timer = null;

state.newline = String.fromCharCode(10);
state.bar = "I am BAR!";    //debug

rpc.exports = {
    init: function (stage, parameters) {
        %(init_code)s
    },
    dispose: function () {
        flush();
    },
```

Note:

- The line `init_session_code += "\n"` is necessary to separate two different JS file by a newline.

- The `init` function seems straightforward to me (below I will show you what is generated at runtime)

- I have inserted the line `state.bar = "I am BAR!";  //debug`.  This will be used to show that `state.bar` is set correctly and shows up in the test handler, while the `state.foo` variable in the --init-session code is NOT set.

For testing I used the following --init-session JS file (entitled `init-session.js`):

```
state.foo = 'I am FOO!';
```

I used the `msvcrt.dll!_lock` function handler for testing:

```
  onEnter: function (log, args, state) {
    log('_lock() [msvcrt.dll]');
    log('state.foo = "' + state.foo + '"')
    log('state.bar = "' + state.bar + '"')
  },
```

And finally, here is the command line I used to test everything:

```
frida-trace -p 17860 -i "msvcrt.dll!_lock" -S "D:\Junk\frida-test\init-session.js"
```

(For the record, I am using the `vlc.exe` application for testing.).

Now we come to the runtime results.  At lines 551-553 I inserted a dump of the string returned by self._create_trace_script(), so we can see exactly what the JS trace looks like:

```
D:\Junk\frida-test>frida-trace -p 17860 -i "msvcrt.dll!_lock" -S "D:\Junk\frida-test\init-session.js"
Instrumenting functions...
----- _create_trace_script()  ------------------------------
'use strict';

var started = Date.now();
var handlers = {};
var state = {};
var pending = [];
var timer = null;

state.newline = String.fromCharCode(10);
state.bar = "I am BAR!";

rpc.exports = {
    init: function (stage, parameters) {
        state.foo = 'I am FOO!';
    },
    dispose: function () {
        flush();
    },
    add: function (targets) {
        targets.forEach(function (target) {
            var h = [parseHandler(target)];
            var name = target.name;
            var targetAddress = target.absolute_address;
            target = null;

            handlers[targetAddress] = h;

            function invokeCallback(callback, context, param) {
                if (callback === undefined)
                    return;

                var timestamp = Date.now() - started;
                var threadId = context.threadId;
                var depth = context.depth;

                function log(message) {
                    emit([timestamp, threadId, depth, targetAddress, message]);
                }

                callback.call(context, log, param, state);
            }

            try {
                Interceptor.attach(ptr(targetAddress), {
                    onEnter: function (args) {
                        invokeCallback(h[0].onEnter, this, args);
                    },
                    onLeave: function (retval) {
                        invokeCallback(h[0].onLeave, this, retval);
                    }
                });
            } catch (e) {
                send({
                    from: '/targets',
                    name: '+error',
                    payload: {
                        message: 'Skipping "' + name + '": ' + e.message
                    }
                });
            }
        });
    },
    update: function (targets) {
        targets.forEach(function (target) {
            handlers[target.absolute_address][0] = parseHandler(target);
        });
    }
};

function emit(event) {
    pending.push(event);

    if (timer === null)
        timer = setTimeout(flush, 50);
}

function flush() {
    if (timer !== null) {
        clearTimeout(timer);
        timer = null;
    }

    if (pending.length === 0)
        return;

    var items = pending;
    pending = [];

    send({
        from: '/events',
        name: '+add',
        payload: {
            items: items
        }
    });
}

function parseHandler(target) {
    try {
        return (1, eval)('(' + target.handler + ')');
    } catch (e) {
        send({
            from: '/targets',
            name: '+error',
            payload: {
                message: 'Invalid handler for "' + target.name + '": ' + e.message
            }
        });
        return {};
    }
}

------------------------------------------------------------
_lock: Loaded handler at "D:\\Junk\\frida-test\\__handlers__\\msvcrt.dll\\_lock.js"
Started tracing 1 function. Press Ctrl+C to stop.
           /* TID 0x42bc */
 12082 ms  _lock() [msvcrt.dll]
 12082 ms  state.foo = "undefined"
 12082 ms  state.bar = "I am BAR!"
 12083 ms  _lock() [msvcrt.dll]
 12083 ms  state.foo = "undefined"
 12083 ms  state.bar = "I am BAR!"
 12083 ms  _lock() [msvcrt.dll]
 12083 ms  state.foo = "undefined"
 12083 ms  state.bar = "I am BAR!"
 ...
 ```

Note:

- The `state.bar = "I am BAR!";` statement is correctly set in the `_lock` handler trace output.

- The `rpc.exports:init()` function does not seem to have been called, as the `state.foo` variable is undefined in the handler trace output.

Summary: Can you tell me why the `rpc.exports:init()` is not being invoked?

Reply:
There isn't any magic that calls `init()`, we need to call it ourselves just like any other RPC method. (It's an application-level protocol that we implement in `frida-gadget` and `frida-inject`.)